ovirt_imageio/client/_upload.py

```diff
@@ -0,0 +1,146 @@
+        default="qcow2",
+        help="Upload image format (default qcow2).")
+
+    cmd.add_argument(
+        "--preallocated",
+        dest="sparse",
+        action="store_false",
+        help="Create preallocated disk. Required when using raw format on "
+             "block based storage domain (iSCSI, FC)")
+
+    cmd.add_argument(
+        "--id", "--disk-id",
+        type=_options.UUID,
+        help="A UUID for the new disk. If not specified oVirt will "
+             "create a new UUID.")
+
+    cmd.add_argument(
+        "filename",
+        type=_options.File,
+        help="Path to image to upload. Supported formats: raw, qcow2, iso.")
+
+
+def upload_disk(args):
+    with _ui.ProgressBar(phase="inspecting image") as progress:
+        disk_info = _probe_image(args.filename, args.format)
+        con = _ovirt.connect(args)
+        with closing(con):
+            progress.phase = "creating disk"
+            disk = _ovirt.add_disk(
+                con=con,
+                name=disk_info.name,
+                initial_size=disk_info.initial_size,
+                provisioned_size=disk_info.provisioned_size,
+                sd_name=args.storage_domain,
+                id=args.id,
+                sparse=args.sparse,
+                enable_backup=disk_info.format == _ovirt.COW,
+                content_type=disk_info.content_type,
+                format=disk_info.format)
+
+            progress.phase = "creating transfer"
+            host = _ovirt.find_host(con, args.storage_domain)
+            transfer = _ovirt.create_transfer(con, disk, host=host)
+            try:
+                progress.phase = "uploading image"
+                _api.upload(
+                    args.filename,
+                    transfer.transfer_url,
+                    args.cafile,
+                    buffer_size=args.buffer_size,
+                    progress=progress,
+                    proxy_url=transfer.proxy_url,
+                    max_workers=args.max_workers)
+            except Exception:
+                progress.phase = "cancelling transfer"
+                _ovirt.cancel_transfer(con, transfer)
+                raise
+
+            progress.phase = "finalizing transfer"
+            _ovirt.finalize_transfer(con, transfer, disk)
+        progress.phase = "upload completed"
+
+
+def _probe_image(filename, dst_format):
+    # Obtain the image info dictionary.
+    img_info = _api.info(filename)
+    if img_info["format"] not in ("qcow2", "raw"):
+        raise RuntimeError(f"Unsupported image format {img_info['format']}")
+    # Obtain all remaining fields for DiskInfo.
+    format = dst_format or img_info["format"]
+    initial_size = _api.measure(filename, format)["required"]
+    content_type = _detect_disk_content_type(
+        img_info["filename"], img_info["format"])
+    name = os.path.splitext(os.path.basename(img_info["filename"]))[0]
+    disk_format = _ovirt.RAW if format == "raw" else _ovirt.COW
```

Comment:
Testing upload ISO:
```
$ ./ovirt-img upload-disk -c engine -s fc-01 /data/tmp/Fedora-Server-dvd-x86_64-36-1.5.iso
[ 100% ] 2.20 GiB, 14.35 s, 156.68 MiB/s | upload completed                    
```

Disk created as qcow2 sparse on block storage.

When starting a vm with this iso, the vm fail to boot from the disk, since oVirt always
uses raw format for cdrom:

```
    <disk type='block' device='cdrom'>
      <driver name='qemu' type='raw' error_policy='report'/>
      <source dev='/rhev/data-center/mnt/blockSD/bafd0f16-9aba-4f9f-ba90-46d3b8a29157/images/830fd34e-1b62-41ed-b03d-b29bff29dded/40a026a6-6a83-46ab-ac6b-b9c6b83e0c91' index='2'>
        <seclabel model='dac' relabel='no'/>
      </source>
      <backingStore/>
      <target dev='sdc' bus='sata'/>
      <readonly/>
      <alias name='ua-4f099cd6-d435-479e-8874-ab4ba3f9c1fd'/>
      <address type='drive' controller='0' bus='0' target='0' unit='2'/>
    </disk>
```

Theoretically we can support qcow2 iso images, but there is no reason to support this.

Reply:
Got it! Thanks for the clarification. Fixed :+1: